File: datastructures/stacks/java/algorithms/MinStack.java
import java.util.Stack;

public class MinStack {

    private Stack<Integer> dataStack = new Stack<>();
    private Stack<Integer> minStack = new Stack<>();

    public MinStack() {
    }

    public void push(int val) {
        dataStack.push(val);
        if (minStack.isEmpty() || val <= minStack.peek()) {
            minStack.push(val);
        }
    }

    public void pop() {
        if (dataStack.peek().equals(minStack.peek())) {
            minStack.pop();
        }
        dataStack.pop();
    }

    public int top() {
        return dataStack.peek();
    }

    public int getMin() {
        return minStack.peek();
    }

    public void printStack() {
        dataStack.stream().forEach(System.out::println);
    }

    public static void main(String[] args) {
        MinStack myStack = new MinStack();

        myStack.push(1);
        myStack.push(2);
        myStack.push(3);
        myStack.push(4);

        System.out.println("==== MY STACK ====");
        myStack.printStack();

        System.out.println("Top of my stack:" + myStack.top());

        myStack.pop();

        System.out.println("==== MY STACK AFTER POP ====");
        myStack.printStack();

        System.out.println("Minimum: " + myStack.getMin());

    }
}

File: datastructures/stacks/python/algorithms/min_stack.py
class MinStack:
    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, x: int) -> None:
        self.stack.append(x)
        if not self.min_stack or x <= self.min_stack[-1]:
            self.min_stack.append(x)

    def pop(self) -> None:
        if self.min_stack[-1] == self.stack[-1]:
            self.min_stack.pop()
        self.stack.pop()

    def top(self) -> int:
        return self.stack[-1]

    def getMin(self) -> int:
        return self.min_stack[-1]

    def print_stack(self):
        self.stack.reverse()
        print(*self.stack)


def main():
    min_stack = MinStack()

    min_stack.push(1)
    min_stack.push(2)
    min_stack.push(3)
    min_stack.push(4)

    print("==== MY STACK ====")
    min_stack.print_stack()

    print("==== MIN VALUE IN THE STACK ====")
    print(min_stack.getMin())


if __name__ == "__main__":
    main()
